controllers/appmesh/cloudmap_controller.go

```diff
@@ -0,0 +1,888 @@
+	ctx := context.Background()
+	log := r.log.WithValues("Virtualnode", req.NamespacedName)
+
+	vNode := &appmesh.VirtualNode{}
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, vNode); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	log.V(2).Info("VNode: ", "awsName: ", vNode.Spec.AWSName,
+		"vNode.PodSelector: ", vNode.Spec.PodSelector)
+	if !vNode.DeletionTimestamp.IsZero() {
+		return r.deleteCloudMapResources(ctx, vNode)
+	}
+	return r.reconcileVirtualNodeWithCloudMap(ctx, vNode)
+}
+
+func (r *cloudMapReconciler) reconcileVirtualNodeWithCloudMap(ctx context.Context, vNode *appmesh.VirtualNode) error {
+
+	if !r.isCloudMapEnabledForVirtualNode(vNode) {
+		return errors.Errorf("CloudMap not enabled for virtualNode: %s", vNode.Name)
+	}
+
+	if !r.isPodSelectorDefinedForVirtualNode(vNode) {
+		return errors.Errorf("PodSelector not defined for virtualNode: %s", vNode.Name)
+	}
+
+	cloudmapConfig := vNode.Spec.ServiceDiscovery.AWSCloudMap
+	creatorRequestID := vNode.UID
+
+	var podsList corev1.PodList
+	var listOptions client.ListOptions
+	listOptions.LabelSelector, _ = metav1.LabelSelectorAsSelector(vNode.Spec.PodSelector)
+	listOptions.Namespace = vNode.Namespace
+
+	if err := r.k8sClient.List(ctx, &podsList, &listOptions); err != nil {
+		r.log.Error(err, "Couldn't retrieve pods for VirtualNode")
+		return err
+	}
+
+	r.log.V(2).Info("Pods found matching this VNode: ", "Total Pods: ", len(podsList.Items))
+	if err := r.finalizerManager.AddFinalizers(ctx, vNode, k8s.FinalizerAWSCloudMapResources); err != nil {
+		r.log.Error(err, "..while adding cloudmap resource finalizer")
+		return err
+	}
+
+	_, err := r.createCloudMapService(ctx, cloudmapConfig, string(creatorRequestID))
+	if err != nil {
+		r.log.Error(err, "failed to create cloudMap Service for", "vNode: ", vNode.Name)
+		return err
+	}
+
+	serviceInstanceSummary := make(map[string]*cloudMapInstanceInfo)
+	for _, pod := range podsList.Items {
+		if pod.DeletionTimestamp != nil {
+			r.log.Info("vNode:", "Pod is being deleted: ", pod.Name)
+			continue
+		}
+
+		instanceId := r.podToInstanceID(&pod)
+		if instanceId == "" {
+			r.log.Info("No IP Address assigned to Pod:", pod.Name, "..Skipping for now")
+			continue
+		}
+
+		instanceInfo := &cloudMapInstanceInfo{}
+		if r.isPodReady(&pod) {
+			instanceInfo.healthStatus = awsservices.InstanceHealthy
+		} else if r.shouldPodBeRegisteredWithCloudMapService(&pod) {
+			instanceInfo.healthStatus = awsservices.InstanceUnHealthy
+		} else {
+			continue
+		}
+		instanceInfo.podName = pod.Name
+		instanceInfo.namespace = pod.Namespace
+		instanceInfo.virtualNodeName = vNode.Name
+		instanceInfo.meshName = vNode.Name
+		instanceInfo.labels = make(map[string]string)
+		for label, v := range pod.Labels {
+			instanceInfo.labels[label] = v
+		}
+		serviceInstanceSummary[instanceId] = instanceInfo
+	}
+
+	if err := r.updateCloudMapServiceInstances(ctx, cloudmapConfig, serviceInstanceSummary, string(creatorRequestID)); err != nil {
```

Comment:
better to have a component dedicated to register/deregister instances from cloudmap

Reply:
done